### PR TITLE
Dynamic Interest Rate with dynamic interest curve

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: contracts/connectors/loantoken/LoanTokenLogicStandard.sol [hunk 1]
```diff
@@ -356,12 +352,7 @@ contract LoanTokenLogicStandard is AdvancedToken {
         view
```
Comment (RomanHiden) [inline contracts/connectors/loantoken/LoanTokenLogicStandard.sol:352]: `tokenPrice` function is public which can be called internally and externally. why having extra _tokenPrice?

File: contracts/modules/ProtocolSettings/ProtocolSettings.sol [hunk 2]
```diff
@@ -136,8 +136,12 @@ contract ProtocolSettings is State, ProtocolSettingsEvents {
                 0x4a99e3a1, // setSwapApprovals(address[])
                 addrs
             );
-            (bool success,) = swapsImpl.delegatecall(data);
-            require(success, "approval calls failed");
+            IDexRecords records = IDexRecords(swapsImpl);
```
Comment (RomanHiden) [inline contracts/modules/ProtocolSettings/ProtocolSettings.sol:148]: T: this is for convenience. however we can just call setSwapApprovals(address[]) on swap implementation directly. do we really need this here?

File: contracts/connectors/loantoken/LoanTokenSettings.sol [hunk 3]
```diff
@@ -106,5 +113,18 @@ contract LoanTokenSettings is AdvancedTokenStorage {
         decimals = IERC20Detailed(loanTokenAddress).decimals();
 
         initialPrice = WEI_PRECISION; // starting price of 1
+
+        IERC20(_loanTokenAddress).approve(bZxContract, uint256(-1));
+    }
+
+    function revokeApproval(
+        address _loanTokenAddress)
+        public
```
Comment (Drypto13) [inline contracts/connectors/loantoken/LoanTokenSettings.sol:124]: needs onlyOwner modifier

File: contracts/modules/ProtocolSettings/ProtocolSettings.sol (hunk 2 above)
Comment (Drypto13) [reply contracts/modules/ProtocolSettings/ProtocolSettings.sol:148]: It is pretty convenient to have and doesn't need to be triggered so I do not see the harm of it

File: contracts/connectors/loantoken/LoanTokenSettings.sol (hunk 3 above)
Comment (RomanHiden) [reply contracts/connectors/loantoken/LoanTokenSettings.sol:124]: I have this concern as well

Comment (RomanHiden) [reply contracts/connectors/loantoken/LoanTokenSettings.sol:124]: has to be guardian

File: contracts/farm/BuyBackAndBurn.sol [hunk 4]
```diff
@@ -22,7 +22,7 @@ contract BuyBackAndBurn is Upgradeable_0_8 {
     address public constant USDC = 0x2791Bca1f2de4661ED88A30C99A7a9449Aa84174;
     address public constant P125 = 0x83000597e8420aD7e9EDD410b2883Df1b83823cF;
     IUniswapV3SwapRouter public constant SWAP_ROUTER =
-        IUniswapV3SwapRouter(0x68b3465833fb72A70ecDF485E0e4C7bD8665Fc45); //uni v3
+        IUniswapV3SwapRouter(0xE592427A0AEce92De3Edee1F18E0157C05861564); //uni v3
```
Comment (RomanHiden) [inline contracts/farm/BuyBackAndBurn.sol:25]: @Drypto13 why do we have other swap router here? and both labeled v3

File: contracts/mixins/InterestHandler.sol [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+ */
+
+pragma solidity 0.5.17;
+
+import "../core/State.sol";
+import "../interfaces/ILoanPool.sol";
+import "../utils/MathUtil.sol";
+
+
+contract InterestHandler is State {
```
Comment (RomanHiden) [inline contracts/mixins/InterestHandler.sol]: @Drypto13 @eugenesh1 try to find a way where we can store improper values interest rate model

File: contracts/farm/BuyBackAndBurn.sol (hunk 4 above)
Comment (Drypto13) [reply contracts/farm/BuyBackAndBurn.sol:25]: it is replacing the swap router with the one that uses deadline on code so it compiles properly

File: contracts/connectors/loantoken/LoanTokenLogicStandard.sol [hunk 6]
```diff
@@ -1024,153 +1000,53 @@ contract LoanTokenLogicStandard is AdvancedToken {
...
         if (assetBorrow != 0 && assetSupply >= assetBorrow) {
-            return _avgBorrowInterestRate(assetBorrow)
+            return nextBorrowRate
                 .mul(_utilizationRate(assetBorrow, assetSupply))
                 .mul(SafeMath.sub(WEI_PERCENT_PRECISION, IBZx(bZxContract).lendingFeePercent()))
                 .div(WEI_PERCENT_PRECISION * WEI_PERCENT_PRECISION);
         }
     }
 
     function _nextBorrowInterestRate(
```
Comment (Drypto13) [inline contracts/connectors/loantoken/LoanTokenLogicStandard.sol:988]: I recommend renaming this to nextBorrowInterestRate and on the interface add it as an overloaded function as the _ prefix should be reserved for internal functions only. I can understand why it makes sense to have both this and the auto filling version as publicly accessible.

File: contracts/swaps/connectors/SwapsImplUniswapV3_ETH.sol [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+                    exactParams[0]
+                );
+            }
+            if (sourceTokenAmountUsed > minSourceTokenAmount) {
+                //does not need safe math as it cannot underflow
+                uint256 displace = _numberAdjustment(
+                    sourceTokenAmountUsed,
+                    minSourceTokenAmount
+                );
+                exactParams[0].amountIn -= displace;
```
Comment (RomanHiden) [inline contracts/swaps/connectors/SwapsImplUniswapV3_ETH.sol]: we don't use math here?

File: contracts/swaps/connectors/SwapsImplUniswapV3_ETH.sol [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+                exactParams[0].amountOut += displace; //adds displacement to first swap set
+                totalAmountsOut += displace;
+            }
+            if (totalAmountsInMax < maxSourceTokenAmount) {
+                //does not need safe math as it cannot overflow
+                uint256 displace = _numberAdjustment(
+                    totalAmountsOut,
+                    maxSourceTokenAmount
+                );
+                exactParams[0].amountInMaximum += displace; //adds displacement to first swap set
```
Comment (RomanHiden) [inline contracts/swaps/connectors/SwapsImplUniswapV3_ETH.sol]: how do you know this won't overflow?

Comment (Drypto13) [reply contracts/swaps/connectors/SwapsImplUniswapV3_ETH.sol]: displace is the difference between maxSourceTokenAmount and totalAmountsInMax and exactParams[0].amountInMaximum < totalAmountsInMax so it should not overflow at any point. I realize now that I used the wrong variable for _numberAdjustment call though.

File: contracts/modules/LoanClosings/LoanClosings_Arbitrum.sol [hunk 9]
```diff
@@ -17,7 +17,6 @@ contract LoanClosings_Arbitrum is LoanClosingsBase_Arbitrum {
         onlyOwner
     {
         _setTarget(this.liquidate.selector, target);
-        _setTarget(this.rollover.selector, target);
         _setTarget(this.closeWithDeposit.selector, target);
         _setTarget(this.closeWithSwap.selector, target);
```
Comment (Drypto13) [inline contracts/modules/LoanClosings/LoanClosings_Arbitrum.sol]: function closeWithSwap() needs to use loanDataBytes for dex selector.

File: contracts/farm/BuyBackAndBurn.sol (hunk 4 above)
Comment (RomanHiden) [reply contracts/farm/BuyBackAndBurn.sol:25]: good

File: contracts/modules/LoanOpenings/LoanOpenings.sol [hunk 10]
```diff
@@ -451,8 +418,8 @@ contract LoanOpenings is State, LoanOpeningsEvents, VaultController, InterestUse
                 loanParamsLocal.collateralToken,                // collateralToken
                 sentValues[1],                                  // newPrincipal
                 sentValues[4],                                  // newCollateral
-                sentValues[0],                                  // interestRate
-                sentValues[2],                                  // interestDuration
+                0,                                              // interestRate (old value: sentValues[0])
```
Comment (RomanHiden) [inline contracts/modules/LoanOpenings/LoanOpenings.sol:405]: T: shouldn't we have event with current interest paying ?

File: contracts/modules/LoanClosings/LoanClosings.sol [hunk 11]
```diff
@@ -17,7 +17,6 @@ contract LoanClosings is LoanClosingsBase {
         onlyOwner
     {
         _setTarget(this.liquidate.selector, target);
```
Comment (RomanHiden) [inline contracts/modules/LoanClosings/LoanClosings.sol:19]: shall we do
`_setTarget(bytes4(keccak256("rollover... ")), address(0));` like above?

File: contracts/modules/LoanClosings/LoanClosingsBase_Arbitrum.sol [hunk 12]
```diff
@@ -777,140 +548,59 @@ contract LoanClosingsBase_Arbitrum is State, LoanClosingsEvents, VaultController
...
-        returns (uint256)
-    {
-        // pay outstanding interest to lender
-        _payInterest(
-            loanLocal.lender,
-            loanParamsLocal.loanToken
-        );
+            // interest is paid before principal
+            if (loanCloseAmount >= loanInterest) {
+                principalAfter = principalBefore - (loanCloseAmount - loanInterest);
```
Comment (RomanHiden) [inline contracts/modules/LoanClosings/LoanClosingsBase_Arbitrum.sol]: T: how do you assure not overflow ? and below as well

Comment (Drypto13) [reply contracts/modules/LoanClosings/LoanClosingsBase_Arbitrum.sol]: for this else statement to trigger, principalBefore + loanInterest is strictly greater than loanCloseAmount so principalBefore is going to be greater than loanCloseAmount - loanInterest. loanCloseAmount >= loanInterest so will not overflow there either.

File: contracts/modules/LoanClosings/LoanClosings.sol (hunk 11 above)
Comment (Drypto13) [reply contracts/modules/LoanClosings/LoanClosings.sol:19]: that should work

File: contracts/swaps/connectors/SwapsImplUniswapV3_ETH.sol [hunk 13]
```diff
@@ -0,0 +1,369 @@
...
+                exactParams[uniqueInputParam].amountIn
+            );
+        }
+        if (totalAmounts < amountIn) {
+            exactParams[0].amountIn = exactParams[0].amountIn.add(
+                amountIn.sub(totalAmounts)
+            ); //adds displacement to first swap set
+        }
+        uint256 tempAmountOut;
+        for (uint256 i = 0; i < exactParams.length; i++) {
```
Comment (RomanHiden) [inline contracts/swaps/connectors/SwapsImplUniswapV3_ETH.sol:108]: this for can be extracted into a function. its used twice

File: contracts/swaps/connectors/SwapsImplUniswapV3_ETH.sol [hunk 14]
```diff
@@ -0,0 +1,369 @@
...
+            );
+            amountOut.add(tempAmountIn);
+        }
+    }
+
+    function _getAmountOut(uint256 amountIn, bytes memory path)
+        public
+        returns (uint256)
+    {
+        uint256 amountOut = IUniswapQuoter(uniswapQuoteContract)
```
Comment (RomanHiden) [inline contracts/swaps/connectors/SwapsImplUniswapV3_ETH.sol]: just return immediately

File: contracts/swaps/connectors/SwapsImplUniswapV3_ETH.sol [hunk 15]
```diff
@@ -0,0 +1,369 @@
...
+            .quoteExactInput(path, amountIn);
+        return amountOut;
+    }
+
+    function _getAmountIn(uint256 amountOut, bytes memory path)
+        public
+        returns (uint256)
+    {
+        uint256 amountIn = IUniswapQuoter(uniswapQuoteContract)
```
Comment (RomanHiden) [inline contracts/swaps/connectors/SwapsImplUniswapV3_ETH.sol]: just return immediately

File: contracts/swaps/connectors/SwapsImplUniswapV3_ETH.sol (hunk 13 above)
Comment (Drypto13) [reply contracts/swaps/connectors/SwapsImplUniswapV3_ETH.sol:108]: different call on the if statements.

File: contracts/connectors/loantoken/LoanTokenLogicStandard.sol (hunk 6 above)
Comment (RomanHiden) [reply contracts/connectors/loantoken/LoanTokenLogicStandard.sol:988]: this is internal but public to allow protocol to call it.

**Review by Drypto13 — APPROVED** [review]
Everything looks good and complete

**Review by RomanHiden — APPROVED** [review]
good

**Pull request merged**